In [1]:
# Import libraries
import pandas as pd

In [2]:
# Read data
wards          = pd.read_pickle('datasets\ward.p')
census         = pd.read_pickle('datasets\census.p')
taxi_owners    = pd.read_pickle(r'datasets\taxi_owners.p')
taxi_veh       = pd.read_pickle(r'datasets\taxi_vehicles.p')
licenses       = pd.read_pickle('datasets\licenses.p')
biz_owners     = pd.read_pickle(r'datasets\business_owners.p')
ridership      = pd.read_pickle('datasets\cta_ridership.p')
cal            = pd.read_pickle('datasets\cta_calendar.p')
stations       = pd.read_pickle('datasets\stations.p')
zip_demo       = pd.read_pickle('datasets\zip_demo.p')
land_use       = pd.read_pickle('datasets\land_use.p')

wards_altered  = pd.read_fwf('datasets\wards_altered.data')
census_altered = pd.read_fwf('datasets\census_altered.data')
wards_altered.drop(labels='Unnamed: 0', axis=1, inplace=True)
census_altered.drop(labels='Unnamed: 0', axis=1, inplace=True)
wards_altered['ward'] = wards_altered.ward.astype(object)

# <font color=darkred>1. Data Merging Basics</font>

Learn how you can merge disparate data using inner joins. By combining information from multiple sources you’ll uncover compelling insights that may have previously been hidden. You’ll also learn how the relationship between those sources, such as one-to-one or one-to-many, can affect your result.

# <font color=darkred>1.1 Inner join</font>

**1. Inner join**
>Welcome! I am Aaren Stubberfield and I will be your instructor for this course. The Pandas package is a powerful tool for manipulating and transforming data in Python. However, when working on an analysis, the data needed could be in multiple tables. This course will focus on the vital skill of merging tables together.

**2. For clarity**
>As we start, two quick clarifications. First, through other courses on DataCamp, you may have learned how to import tabular data as DataFrames. In this course, you may hear the words table and DataFrame, but they are equivalent here. Second, we will refer to combining different tables together as merging tables, but note that some refer to this same process as joining.
>(1) Photo by David Travis on Unsplash

**3. Chicago data portal dataset**
>To help us learn about merging tables, we will use data from the city of Chicago data portal.
>(1) Photo by Pedro Lastra on Unsplash

**4. Datasets for example**
>The city of Chicago is divided into fifty local neighborhoods called wards. We have a table with data about the local government offices in each ward. In this example, we want to merge the local government data with census data about the population of each ward.
>(1) Ward image By Alissapump, Own work, CC BY-SA 3.0

**5. The ward data**
>If we look at the wards table, we have information about the local government of each ward, such as the government office address. This table has 50 rows and 4 columns, or one row for each ward.

**6. Census data**
>The census table contains the population of each ward in 2000 and 2010, and that change as a percentage. Additionally, it includes the address for the center of each ward. This table has 50 rows and 6 columns.

**7. Merging tables**
>The two tables are related by their ward column. We can merge them together, matching the ward number from each row of the wards table to the ward numbers from the census table. For example, the second ward in the wards table with Alderman Brian Hopkins would be matched with row 2 of the census table where the population in 2000 was 54,361.

**8. Inner join**
>The pandas package has an excellent DataFrame method for performing this type of merge called merge. The merge method takes the first DataFrame, wards, and merges it with the second DataFrame, census. We use the on argument to tell the method that we want to merge the two DataFrames on the ward column. Since we listed the wards table first, its columns will appear first in the output, followed by the columns from the census table. In this example, the merge returns a DataFrame with 50 rows and 9 columns, where the returned rows have matching values for the ward column in both tables. This is called an inner join.

**9. Inner join**
>An inner join will only return rows that have matching values in both tables.

**10. Suffixes**
>You may have noticed that the merged table has columns with suffixes of underscore x or y. This is because both the wards and census tables contained address and zip columns. To avoid multiple columns with the same name, they are automatically given a suffix by the merge method.

**11. Suffixes**
>We can use the suffix argument of the merge method to control this behavior. We provide a tuple where all of the overlapping columns in the left table are given the suffix '_ward', and those of the right table will be given the suffix '_cen'. This makes it easier for us to tell the difference between the columns.

**12. Let's practice!**
>Now let's practice using the merge method.

In [3]:
# The ward data
print(wards.shape)
wards.head()

(50, 4)


ward            alderman                          address    zip
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649

In [4]:
# Census data
print(census.shape)
census.head()

(50, 6)


ward  pop_2000  pop_2010 change                                  address  \
0    1     52951     56149     6%              2765 WEST SAINT MARY STREET   
1    2     54361     55805     3%                 WM WASTE MANAGEMENT 1500   
2    3     40385     53039    31%                      17 EAST 38TH STREET   
3    4     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   
4    5     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   

     zip  
0  60647  
1  60622  
2  60653  
3  60653  
4  60637

In [5]:
wards_census = wards.merge(census, on='ward')
print(wards_census.shape)
wards_census.head(4)

(50, 9)


ward            alderman                        address_x  zip_x  pop_2000  \
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647     52951   
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622     54361   
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609     40385   
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616     51953   

   pop_2010 change                                address_y  zip_y  
0     56149     6%              2765 WEST SAINT MARY STREET  60647  
1     55805     3%                 WM WASTE MANAGEMENT 1500  60622  
2     53039    31%                      17 EAST 38TH STREET  60653  
3     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL  60653

In [6]:
print(wards_census.columns)

Index(['ward', 'alderman', 'address_x', 'zip_x', 'pop_2000', 'pop_2010',
       'change', 'address_y', 'zip_y'],
      dtype='object')


In [7]:
wards_census = wards.merge(census, on='ward', suffixes=('_ward','_cen'))
print(wards_census.head())
print(wards_census.shape)

  ward            alderman                     address_ward zip_ward  \
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE    60647   
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE    60622   
2    3          Pat Dowell          5046 SOUTH STATE STREET    60609   
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR    60616   
4    5  Leslie A. Hairston            2325 EAST 71ST STREET    60649   

   pop_2000  pop_2010 change                              address_cen zip_cen  
0     52951     56149     6%              2765 WEST SAINT MARY STREET   60647  
1     54361     55805     3%                 WM WASTE MANAGEMENT 1500   60622  
2     40385     53039    31%                      17 EAST 38TH STREET   60653  
3     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   60653  
4     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   60637  
(50, 9)


In [8]:
# Merge the taxi_owners and taxi_veh tables setting a suffix
taxi_own_veh = taxi_owners.merge(taxi_veh, on='vid', suffixes=('_own', '_veh'))

# Print the column names of taxi_own_veh
print(taxi_own_veh.columns)

Index(['rid', 'vid', 'owner_own', 'address', 'zip', 'make', 'model', 'year',
       'fuel_type', 'owner_veh'],
      dtype='object')


# <font color=darkred>1.2 What column to merge on?</font>

Chicago provides a list of taxicab owners and vehicles licensed to operate within the city, for public safety. Your goal is to merge two tables together. One table is called taxi_owners, with info about the taxi cab company owners, and one is called taxi_veh, with info about each taxi cab vehicle. Both the taxi_owners and taxi_veh tables have been loaded for you and you can explore them in the console.

**Instructions**

Choose the column you would use to merge the two tables on using the .merge() method.

**Possible Answers**
- on='rid'
- <u><font color=red>on='vid'</font></u>
- on='year'
- on='zip'

**Results**

<font color=darkgreen>Yes, great job! Both DataFrames contained the column vid. Now continue on to the next exercise where you will using this information to merge the tables.</font>

In [9]:
taxi_owners.head()

rid   vid           owner                 address    zip
0  T6285  6285  AGEAN TAXI LLC     4536 N. ELSTON AVE.  60630
1  T4862  4862    MANGIB CORP.  5717 N. WASHTENAW AVE.  60659
2  T1495  1495   FUNRIDE, INC.     3351 W. ADDISON ST.  60618
3  T4231  4231    ALQUSH CORP.   6611 N. CAMPBELL AVE.  60645
4  T5971  5971  EUNIFFORD INC.     3351 W. ADDISON ST.  60618

In [10]:
taxi_veh.head()

vid    make   model  year fuel_type                owner
0  2767  TOYOTA   CAMRY  2013    HYBRID       SEYED M. BADRI
1  1411  TOYOTA    RAV4  2017    HYBRID          DESZY CORP.
2  6500  NISSAN  SENTRA  2019  GASOLINE       AGAPH CAB CORP
3  2746  TOYOTA   CAMRY  2013    HYBRID  MIDWEST CAB CO, INC
4  5922  TOYOTA   CAMRY  2013    HYBRID       SUMETTI CAB CO

## <font color=darkred>1.3 Your first inner join</font>

You have been tasked with figuring out what the most popular types of fuel used in Chicago taxis are. To complete the analysis, you need to merge the taxi_owners and taxi_veh tables together on the vid column. You can then use the merged table along with the .value_counts() method to find the most common fuel_type.

Since you'll be working with pandas throughout the course, the package will be preloaded for you as pd in each exercise in this course. Also the taxi_owners and taxi_veh DataFrames are loaded for you.

**Instructions**

1. Merge taxi_owners with taxi_veh on the column vid, and save the result to taxi_own_veh.
2. Set the left and right table suffixes for overlapping columns of the merge to _own and _veh, respectively.
3. Select the fuel_type column from taxi_own_veh and print the value_counts() to find the most popular fuel_types used.

**Results**

<font color=darkgreen>Bravo! You correctly merged the two tables together and found out that the most common fuel type for taxis in Chicago are hybrids.</font>

In [11]:
# Merge the taxi_owners and taxi_veh tables
taxi_own_veh = taxi_owners.merge(taxi_veh, on='vid')

# Print the column names of the taxi_own_veh
print(taxi_own_veh.columns)

Index(['rid', 'vid', 'owner_x', 'address', 'zip', 'make', 'model', 'year',
       'fuel_type', 'owner_y'],
      dtype='object')


In [12]:
# Merge the taxi_owners and taxi_veh tables setting a suffix
taxi_own_veh = taxi_owners.merge(taxi_veh, on='vid', suffixes=('_own', '_veh'))

# Print the column names of taxi_own_veh
print(taxi_own_veh.columns)

Index(['rid', 'vid', 'owner_own', 'address', 'zip', 'make', 'model', 'year',
       'fuel_type', 'owner_veh'],
      dtype='object')


In [13]:
# Print the value_counts to find the most popular fuel_type
print(taxi_own_veh['fuel_type'].value_counts())

HYBRID                    2792
GASOLINE                   611
FLEX FUEL                   89
COMPRESSED NATURAL GAS      27
Name: fuel_type, dtype: int64


## <font color=darkred>1.4 Inner joins and number of rows returned</font>

All of the merges you have studied to this point are called inner joins. It is necessary to understand that inner joins only return the rows with matching values in both tables. You will explore this further by reviewing the merge between the wards and census tables, then comparing it to merges of copies of these tables that are slightly altered, named wards_altered, and census_altered. The first row of the wards column has been changed in the altered tables. You will examine how this affects the merge between them. The tables have been loaded for you.

For this exercise, it is important to know that the wards and census tables start with 50 rows.

**Instructions**

1. Merge wards and census on the ward column and save the result to wards_census.
2. Merge the wards_altered and census tables on the ward column, and notice the difference in returned rows.
3. Merge the wards and census_altered tables on the ward column, and notice the difference in returned rows.

**Results**

<font color=darkgreen>Great job! In step 1, the .merge() returned a table with the same number of rows as the original wards table. However, in steps 2 and 3, using the altered tables with the altered first row of the ward column, the number of returned rows was fewer. There was not a matching value in the ward column of the other table. Remember that .merge() only returns rows where the values match in both tables.</font>

In [14]:
wards_altered.head()

ward            alderman                          address    zip
0   61  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649

In [15]:
census_altered.head()

ward  pop_2000  pop_2010 change                                  address  \
0  None     52951     56149     6%              2765 WEST SAINT MARY STREET   
1     2     54361     55805     3%                 WM WASTE MANAGEMENT 1500   
2     3     40385     53039    31%                      17 EAST 38TH STREET   
3     4     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   
4     5     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   

     zip  
0  60647  
1  60622  
2  60653  
3  60653  
4  60637

In [16]:
# Merge the wards and census tables on the ward column
wards_census = wards.merge(census, on='ward')

# Print the shape of wards_census
print('wards_census table shape:', wards_census.shape)

wards_census table shape: (50, 9)


In [17]:
# Print the first few rows of the wards_altered table to view the change 
print(wards_altered[['ward']].head())

# Merge the wards_altered and census tables on the ward column
wards_altered_census = wards_altered.merge(census, on='ward')

# Print the shape of wards_altered_census
print('wards_altered_census table shape:', wards_altered_census.shape)

  ward
0   61
1    2
2    3
3    4
4    5
wards_altered_census table shape: (0, 9)


In [18]:
# Print the first few rows of the census_altered table to view the change 
print(census_altered[['ward']].head())

# Merge the wards and census_altered tables on the ward column
wards_census_altered = wards.merge(census_altered, on='ward')

# Print the shape of wards_census_altered
print('wards_census_altered table shape:', wards_census_altered.shape)

   ward
0  None
1     2
2     3
3     4
4     5
wards_census_altered table shape: (49, 9)


## <font color=darkred>1.5 One-to-many relationships</font>

**1. One to many relationships**
>Welcome back! In the last lesson, we learned how to merge two DataFrames together with the merge method. In this lesson, we'll discuss different types of relationships between tables. In particular, we will discuss the one-to-many relationship. But first, let's quickly consider what a one-to-one relationship is.

**2. One-to-one**
>In a one-to-one relationship, every row in the left table is related to one and only one row in the right table.

**3. One-to-one example**
>We looked at a one-to-one relationship earlier. Recall the relationship between the wards table and the census table. Every row in the wards table is related to only one row in the census table, so there is only one row for ward 3 in each table. Practically speaking, it only makes sense that there is one row of population information for each ward. It wouldn't make sense if the census table contained multiple population values in 2000 for the third ward.

**4. One-to-many**
>So, what is a one-to-many relationship? Well, in a one-to-many relationship, every row in the left table is related to one or more rows in the right table.

**5. One-to-many example**
>To provide an example of a one-to-many relationship, let's think back to our wards table. Within each ward, there are many businesses. We will merge the wards table with a table of licensed businesses in each ward.

**6. One-to-many example**
>The business license data is stored in another table called licenses. It holds info such as the business address and ward the business is located within.

**7. One-to-many example**
>The two DataFrames are related to each other by their ward column.

**8. One-to-many example**
>When we merge the two tables together with the merge method, setting the 'on' attribute to the column ward, the resulting table has both local ward data and business license data. Notice that ward 1 and its alderman Joe is repeated in the resulting table because the licenses table has many businesses in the 1st ward. Pandas takes care of the one-to-many relationships for us and doesn't require anything special on our end. We can use the same syntax as we did with one-to-one relationships.

**9. One-to-many example**
>By printing the shape, we can see that our original wards table has 50 rows. After merging the wards table with the licenses table, the resulting table has 10,000 rows. When you merge tables that have a one-to-many relationship, the number of rows returned will likely be different than the number in the left table.

**10. Let's practice!**
>Now let's make the one-to-many relationship idea more concrete by practicing.

In [19]:
print(licenses.shape)
licenses.head()

(10000, 6)


account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613

In [20]:
ward_licenses = wards.merge(licenses, on='ward', suffixes=('_ward','_lic'))
ward_licenses.head()

ward            alderman               address_ward zip_ward account  aid  \
0    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE    60647   12024  NaN   
1    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE    60647   14446  743   
2    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE    60647   14624  775   
3    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE    60647   14987  NaN   
4    1  Proco "Joe" Moreno  2058 NORTH WESTERN AVENUE    60647   15642  814   

               business              address_lic zip_lic  
0   DIGILOG ELECTRONICS       1038 N ASHLAND AVE   60622  
1      EMPTY BOTTLE INC   1035 N WESTERN AVE 1ST   60622  
2  LITTLE MEL'S HOT DOG    2205 N CALIFORNIA AVE   60647  
3    MR. BROWN'S LOUNGE   2301 W CHICAGO AVE 1ST   60622  
4          Beat Kitchen  2000-2100 W DIVISION ST   60622

In [21]:
print(wards.shape)
print(ward_licenses.shape)

(50, 4)
(10000, 9)


## <font color=darkred>1.6 One-to-many classification</font>

Understanding the difference between a one-to-one and one-to-many relationship is a useful skill. In this exercise, consider a set of tables from an e-commerce website. The hypothetical tables are the following:

- A customer table with information about each customer
- A cust_tax_info table with customers unique tax IDs
- An orders table with information about each order
- A products table with details about each unique product sold
- An inventory table with information on how much total inventory is available to sell for each product

**Instructions**
1. Select the relationship type that is most appropriate for the relationship between the different tables: One-to-one, or One-to-many.

**Possible Answers**

One-to-one
- The relationship between customer and cust_tax_info
- The relationship between products and inventory

One-to-many
- The relationship between customer and orders
- The relationship between products and customer

**Results**

<font color=darkgreen>Great job on classifying the relationships into the right categories. Knowing if a relationship should be a one-to-one or one-to-many will allow you to troubleshoot a merge that appears incorrect.</font>

## <font color=darkred>1.7 One-to-many merge</font>

A business may have one or multiple owners. In this exercise, you will continue to gain experience with one-to-many merges by merging a table of business owners, called biz_owners, to the licenses table. Recall from the video lesson, with a one-to-many relationship, a row in the left table may be repeated if it is related to multiple rows in the right table. In this lesson, you will explore this further by finding out what is the most common business owner title. (i.e., secretary, CEO, or vice president)

The licenses and biz_owners DataFrames are loaded for you.

**Instructions**

1. Starting with the licenses table on the left, merge it to the biz_owners table on the column account, and save the results to a variable named licenses_owners.
2. Group licenses_owners by title and count the number of accounts for each title. Save the result as counted_df
3. Sort counted_df by the number of accounts in descending order, and save this as a variable named sorted_df.
4. Use the .head() method to print the first few rows of the sorted_df.

**Results**

<font color=darkgreen>Wonderful! After merging the tables together, you counted the number of repeated rows with the combination of .groupby() and .agg() statements. You see that president, followed by secretary, are the most common business owner titles.</font>

In [22]:
print(biz_owners.shape)
biz_owners.head()

(21352, 4)


account first_name  last_name      title
0      10      PEARL    SHERMAN  PRESIDENT
1      10      PEARL    SHERMAN  SECRETARY
2   10002     WALTER     MROZEK    PARTNER
3   10002     CELINA     BYRDAK    PARTNER
4   10005      IRENE  ROSENFELD  PRESIDENT

In [23]:
print(licenses.shape)
licenses.head()

(10000, 6)


account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613

In [24]:
# Merge the licenses and biz_owners table on account
licenses_owners = licenses.merge(biz_owners, on='account')

# Group the results by title then count the number of accounts
counted_df = licenses_owners.groupby('title').agg({'account':'count'})

# Sort the counted_df in desending order
sorted_df = counted_df.sort_values('account', ascending=False)

# Use .head() method to print the first few rows of sorted_df
sorted_df.head()

account
title                   
PRESIDENT           6259
SECRETARY           5205
SOLE PROPRIETOR     1658
OTHER               1200
VICE PRESIDENT       970

## <font color=darkred>1.8 Merging multiple DataFrames</font>

**1. Merging multiple DataFrames**
>Welcome back. In our last lesson, we learned how to merge two tables with a one-to-many relationship using the merge method. Merging data like this is a necessary skill to bring together data from different sources to answer some more complex data questions.

**2. Merging multiple tables**
>Sometimes we need to merge together more than just two tables to complete our analysis.

**3. Remembering the licenses table**
>In the previous lesson, we used two tables from the city of Chicago. One table contained business licenses issued by the city.

**4. Remembering the wards table**
>The other table listed info about the local neighborhoods called wards, including the local government official's office.

**5. Review new data**
>Now, we also have a table of businesses that have received small business grant money from Chicago. The grants are funded by taxpayer money. Therefore, it would be helpful to analyze how much grant money each business received and in what ward that business is located. We then could determine if one ward's businesses received a disproportionately large amount of grant money.

**6. Tables to merge**
>To pull all of this information together, let's first connect our grants table to our licenses table. The two tables are related by their company name and location. Let's pause here for a moment.

**7. Theoretical merge**
>If we merge the two tables only using the zip column, then the 60616 zip of Reggie's bar from the licenses table will be matched to multiple businesses in the grants table with the same zip. Our code sample prints the first few rows and some columns of the merged table. The output of the merge duplicates Reggie's bar for each matching zip in the grants table, which is not what we want. If instead, we merged on address only, there's a small risk that the address would repeat in different parts of the city. Therefore, the best option is to merge the tables using the combination of both address and zip code.

**8. Single merge**
>We merge the two DataFrames as shown before, except in this case, we pass a list of the column names we want to merge on to the 'on' argument. This allows us to use multiple columns in the merge. As before, the matching rows between the two DataFrames are returned with the columns from the grant table listed first. However, when we merge on two columns, in this case address and zip code, we are requiring that both the address and zip code of a row in the left table match the address and zip code of a row in the right table in order for them to be linked to each other in the merge.

**9. Merging multiple tables**
>We can now extend this example to a third table. First, we merge the grants table with the wards table on the ward column again, adding suffixes to the repeated column names. Note that we're using Python's backslash line continuation method to add the second merge on the next line. Python will read this as just one line of code. Without this, Python will throw a syntax error since it will parse it as two separate lines of code, so don't forget your backslash. Now our output table has information about grants, business, and wards. We can now complete our analysis.

**10. Results**
>We can now sum the grants by ward and plot the results. Some wards have received more grants than others.

**11. Merging even more...**
>We could continue to merge additional tables as needed. We stopped at three, but if needed, we could continue to add more. The code here shows the pattern you would follow as you merge more tables.

**12. Let's practice!**
>Now, let's practice merging multiple tables.

## <font color=darkred>1.9 Total riders in a month</font> 

Your goal is to find the total number of rides provided to passengers passing through the Wilson station (station_name == 'Wilson') when riding Chicago's public transportation system on weekdays (day_type == 'Weekday') in July (month == 7). Luckily, Chicago provides this detailed data, but it is in three different tables. You will work on merging these tables together to answer the question. This data is different from the business related data you have seen so far, but all the information you need to answer the question is below.

The cal, ridership, and stations DataFrames have been loaded for you. The relationship between the tables can be seen in the diagram below.

<img src="images/cta_L_diagram.png" width="50%"/>

**Instructions**

1. Merge the ridership and cal tables together, starting with the ridership table on the left and save the result to the variable ridership_cal. If you code takes too long to run, your merge conditions might be incorrect.
2. Extend the previous merge to three tables by also merging the stations table.
3. Create a variable called filter_criteria to select the appropriate rows from the merged table so that you can sum the rides column.


**Results**

<font color=darkgreen>Awesome work! You merged three DataFrames together, including merging two tables on multiple columns. Once the tables were merged, you filtered and selected just like any other DataFrame. Finally, you found out that the Wilson station had 140,005 riders during weekdays in July.</font>

In [25]:
print(cal.head())
print(ridership.head())
print(stations.head())

   year  month  day        day_type
0  2019      1    1  Sunday/Holiday
1  2019      1    2         Weekday
2  2019      1    3         Weekday
3  2019      1    4         Weekday
4  2019      1    5        Saturday
  station_id  year  month  day  rides
0      40010  2019      1    1    576
1      40010  2019      1    2   1457
2      40010  2019      1    3   1543
3      40010  2019      1    4   1621
4      40010  2019      1    5    719
  station_id        station_name                 location
0      40010  Austin-Forest Park  (41.870851, -87.776812)
1      40020         Harlem-Lake  (41.886848, -87.803176)
2      40030        Pulaski-Lake  (41.885412, -87.725404)
3      40040        Quincy/Wells   (41.878723, -87.63374)
4      40050               Davis   (42.04771, -87.683543)


In [26]:
# Merge the ridership and cal tables
ridership_cal = ridership.merge(cal, on=['year', 'month', 'day'])
ridership_cal.head()

station_id  year  month  day  rides        day_type
0      40010  2019      1    1    576  Sunday/Holiday
1      40080  2019      1    1   1839  Sunday/Holiday
2      40770  2019      1    1   2724  Sunday/Holiday
3      40120  2019      1    1    754  Sunday/Holiday
4      40540  2019      1    1   2175  Sunday/Holiday

In [27]:
# Merge the ridership, cal, and stations tables
ridership_cal_stations = ridership.merge(cal, on=['year','month','day']) \
                                  .merge(stations, on='station_id')
ridership_cal_stations.head()

station_id  year  month  day  rides        day_type        station_name  \
0      40010  2019      1    1    576  Sunday/Holiday  Austin-Forest Park   
1      40010  2019      1    2   1457         Weekday  Austin-Forest Park   
2      40010  2019      1    3   1543         Weekday  Austin-Forest Park   
3      40010  2019      1    4   1621         Weekday  Austin-Forest Park   
4      40010  2019      1    5    719        Saturday  Austin-Forest Park   

                  location  
0  (41.870851, -87.776812)  
1  (41.870851, -87.776812)  
2  (41.870851, -87.776812)  
3  (41.870851, -87.776812)  
4  (41.870851, -87.776812)

In [28]:
# Create a filter to filter ridership_cal_stations
filter_criteria = ((ridership_cal_stations['month'] == 7) 
                   & (ridership_cal_stations['day_type'] == 'Weekday') 
                   & (ridership_cal_stations['station_name'] == 'Wilson'))

# Use .loc and the filter to select for rides
print(ridership_cal_stations.loc[filter_criteria, 'rides'].sum())

140005


## <font color=darkred>1.10 Three table merge</font>

To solidify the concept of a three DataFrame merge, practice another exercise. A reasonable extension of our review of Chicago business data would include looking at demographics information about the neighborhoods where the businesses are. A table with the median income by zip code has been provided to you. You will merge the licenses and wards tables with this new income-by-zip-code table called zip_demo.

The licenses, wards, and zip_demo DataFrames have been loaded for you.

**Instructions**

1. Starting with the licenses table, merge to it the zip_demo table on the zip column. Then merge the resulting table to the wards table on the ward column. Save result of the three merged tables to a variable named licenses_zip_ward.
2. Group the results of the three merged tables by the column alderman and find the median income.

**Results**

<font color=darkgreen>Nice work! You successfully merged three tables together. With the merged data, you can complete your income analysis. You see that only a few aldermen represent businesses in areas where the median income is greater than $62,000, which is the median income for the state of Illinois.</font>

In [29]:
print(licenses.head())
print(wards.head())
print(zip_demo.head())

  account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S BAR & GRILL       2105 S STATE ST  60616
1      10   10  829                 HONEYBEERS   13200 S HOUSTON AVE  60633
2   10002   14  775                CELINA DELI     5089 S ARCHER AVE  60632
3   10005   12  NaN  KRAFT FOODS NORTH AMERICA        2005 W 43RD ST  60609
4   10044   44  638  NEYBOUR'S TAVERN & GRILLE  3651 N SOUTHPORT AVE  60613
  ward            alderman                          address    zip
0    1  Proco "Joe" Moreno        2058 NORTH WESTERN AVENUE  60647
1    2       Brian Hopkins       1400 NORTH  ASHLAND AVENUE  60622
2    3          Pat Dowell          5046 SOUTH STATE STREET  60609
3    4    William D. Burns  435 EAST 35TH STREET, 1ST FLOOR  60616
4    5  Leslie A. Hairston            2325 EAST 71ST STREET  60649
     zip  income
0  60630   70122
1  60640   50488
2  60622   87143
3  60614  100116
4  60608   41226


In [30]:
# Merge licenses and zip_demo, on zip; and merge the wards on ward
licenses_zip_ward = licenses.merge(zip_demo, on='zip') \
                            .merge(wards, on='ward')
print(licenses_zip_ward.head())

# Print the results by alderman and show median income
licenses_zip_ward.groupby('alderman').agg({'income':'median'})

  account ward  aid               business                address_x  zip_x  \
0  307071    3  743   REGGIE'S BAR & GRILL          2105 S STATE ST  60616   
1   11280    3  763              PRIME WAY      2251 S STATE ST 1ST  60616   
2   15015    3  NaN  SOUTHVIEW MANOR, INC.      3311 S MICHIGAN AVE  60616   
3   19168    3  666               BP AMOCO  3101 S MICHIGAN AVE 1ST  60616   
4  205980    3  763  J & J  FISH & CHICKEN            8 E CERMAK RD  60616   

   income    alderman                address_y  zip_y  
0   46340  Pat Dowell  5046 SOUTH STATE STREET  60609  
1   46340  Pat Dowell  5046 SOUTH STATE STREET  60609  
2   46340  Pat Dowell  5046 SOUTH STATE STREET  60609  
3   46340  Pat Dowell  5046 SOUTH STATE STREET  60609  
4   46340  Pat Dowell  5046 SOUTH STATE STREET  60609  


income
alderman                         
Ameya Pawar                 66246
Anthony A. Beale            38206
Anthony V. Napolitano       82226
Ariel E. Reyboras           41307
Brendan Reilly             110215
Brian Hopkins               87143
Carlos Ramirez-Rosa         66246
Carrie M. Austin            38206
Chris Taliaferro            55566
Daniel "Danny" Solis        41226
David H. Moore              33304
Deborah Mell                66246
Debra L. Silverstein        50554
Derrick G. Curtis           65770
Edward M. Burke             42335
Emma M. Mitts               36283
George Cardenas             33959
Gilbert Villegas            41307
Gregory I. Mitchell         24941
Harry Osterman              45442
Howard B. Brookins, Jr.     33304
James Cappleman             79565
Jason C. Ervin              41226
Joe Moore                   39163
John S. Arena               70122
Leslie A. Hairston          28024
Margaret Laurino            70122
Marty Quinn                 67045
Matthew J. O'Shea           59488
Michael R. Zalewski         42335
Michael Scott, Jr.          31445
Michelle A. Harris          32558
Michelle Smith             100116
Milagros "Milly" Santiago   41307
Nicholas Sposato            62223
Pat Dowell                  46340
Patrick Daley Thompson      41226
Patrick J. O'Connor         50554
Proco "Joe" Moreno          87143
Raymond A. Lopez            33959
Ricardo Munoz               31445
Roberto Maldonado           68223
Roderick T. Sawyer          32558
Scott Waguespack            68223
Susan Sadlowski Garza       38417
Tom Tunney                  88708
Toni L. Foulkes             27573
Walter Burnett, Jr.         87143
William D. Burns           107811
Willie B. Cochran           28024

## <font color=darkred>1.11 One-to-many merge with multiple tables</font>

In this exercise, assume that you are looking to start a business in the city of Chicago. Your perfect idea is to start a company that uses goats to mow the lawn for other businesses. However, you have to choose a location in the city to put your goat farm. You need a location with a great deal of space and relatively few businesses and people around to avoid complaints about the smell. You will need to merge three tables to help you choose your location. The land_use table has info on the percentage of vacant land by city ward. The census table has population by ward, and the licenses table lists businesses by ward.

The land_use, census, and licenses tables have been loaded for you.

**Instructions**

1. Merge land_use and census on the ward column. Merge the result of this with licenses on the ward column, using the suffix _cen for the left table and _lic for the right table. Save this to the variable land_cen_lic
2. Group land_cen_lic by ward, pop_2010 (the population in 2010), and vacant, then count the number of accounts. Save the results to pop_vac_lic.
3. Sort pop_vac_lic by vacant, account, andpop_2010 in descending, ascending, and ascending order respectively. Save it as sorted_pop_vac_lic.


**Results**

<font color=darkgreen>Great job putting your new skills into action. You merged multiple tables with varying relationships and added suffixes to make your column names clearer. Using your skills, you were able to pull together information from different tables to see that the 7th ward would be a good place to build your goat farm!</font>

In [31]:
print(land_use.head())
print(census.head())
print(licenses.head())

  ward  residential  commercial  industrial  vacant  other
0    1           41           9           2       2     46
1    2           31          11           6       2     50
2    3           20           5           3      13     59
3    4           22          13           0       7     58
4    5           25           3           1       3     68
  ward  pop_2000  pop_2010 change                                  address  \
0    1     52951     56149     6%              2765 WEST SAINT MARY STREET   
1    2     54361     55805     3%                 WM WASTE MANAGEMENT 1500   
2    3     40385     53039    31%                      17 EAST 38TH STREET   
3    4     51953     54589     5%  31ST ST HARBOR BUILDING LAKEFRONT TRAIL   
4    5     55302     51455    -7%  JACKSON PARK LAGOON SOUTH CORNELL DRIVE   

     zip  
0  60647  
1  60622  
2  60653  
3  60653  
4  60637  
  account ward  aid                   business               address    zip
0  307071    3  743       REGGIE'S 

In [32]:
# Merge land_use and census and merge result with licenses including suffixes
land_cen_lic = land_use.merge(census, on='ward') \
                       .merge(licenses, on='ward', suffixes=('_cen', '_lic'))
land_cen_lic.head()

ward  residential  commercial  industrial  vacant  other  pop_2000  \
0    1           41           9           2       2     46     52951   
1    1           41           9           2       2     46     52951   
2    1           41           9           2       2     46     52951   
3    1           41           9           2       2     46     52951   
4    1           41           9           2       2     46     52951   

   pop_2010 change                  address_cen zip_cen account  aid  \
0     56149     6%  2765 WEST SAINT MARY STREET   60647   12024  NaN   
1     56149     6%  2765 WEST SAINT MARY STREET   60647   14446  743   
2     56149     6%  2765 WEST SAINT MARY STREET   60647   14624  775   
3     56149     6%  2765 WEST SAINT MARY STREET   60647   14987  NaN   
4     56149     6%  2765 WEST SAINT MARY STREET   60647   15642  814   

               business              address_lic zip_lic  
0   DIGILOG ELECTRONICS       1038 N ASHLAND AVE   60622  
1      EMPTY BOTTLE INC   1035 N WESTERN AVE 1ST   60622  
2  LITTLE MEL'S HOT DOG    2205 N CALIFORNIA AVE   60647  
3    MR. BROWN'S LOUNGE   2301 W CHICAGO AVE 1ST   60622  
4          Beat Kitchen  2000-2100 W DIVISION ST   60622

In [33]:
# Merge land_use and census and merge result with licenses including suffixes
land_cen_lic = land_use.merge(census, on='ward') \
                    .merge(licenses, on='ward', suffixes=('_cen','_lic'))

# Group by ward, pop_2010, and vacant, then count the # of accounts
pop_vac_lic = land_cen_lic.groupby(['ward', 'pop_2010', 'vacant'], 
                                   as_index=False).agg({'account':'count'})
pop_vac_lic.head()

ward  pop_2010  vacant  account
0    1     56149       2      253
1   10     51535      14      130
2   11     51497       5      201
3   12     52235       4      255
4   13     53722       1      101

In [34]:
# Sort pop_vac_lic and print the results
sorted_pop_vac_lic = pop_vac_lic.sort_values(['vacant', 'account', 'pop_2010'], 
                                             ascending=[False, True, True])

# Print the top few rows of sorted_pop_vac_lic
sorted_pop_vac_lic.head()

ward  pop_2010  vacant  account
47    7     51581      19       80
12   20     52372      15      123
1    10     51535      14      130
16   24     54909      13       98
7    16     51954      13      156

# Aditional material

- Datacamp course: https://learn.datacamp.com/courses/joining-data-with-pandas